### ISSUE TITLE
SMILES conversion to SDF with --gen2D fails on arm64 architecture

### ISSUE THREAD

**Issue by merkys** [openbabel/openbabel#2417]
I believe this to be a bug with Open Babel. To me it seems as resurfacing of #1851 on arm64 architecture.

## Environment Information
Open Babel version: 3.1.1+dfsg-6, as packaged in Debian
Operating system and version: Debian unstable a.k.a. sid

## Expected Behavior
* The following results in a valid SDF file:
```bash
$ echo 'O=C(C1=CN=CS1)N1C2CCC1CN(CC1CC3CCC1O3)C2' | obabel -i smi -o sdf --gen2D
```
* The following test passes:
```bash
$ python3 test/testbindings.py TestSuite.testTemplates
```

## Actual Behavior
```bash
$ echo 'O=C(C1=CN=CS1)N1C2CCC1CN(CC1CC3CCC1O3)C2' | obabel -i smi -o sdf --gen2D
==============================
*** Open Babel Error  in Convert
  Convert failed with an exception
0 molecules converted
$ python3 test/testbindings.py TestSuite.testTemplates
testTemplates (__main__.TestSuite)
Check for regressions to #1851 ... terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 16) >= this->size() (which is 8)
Aborted
```
The failure is observed on arm64 architecture. Tested also on amd64 and mipsel, both worked fine.

**Comment by merkys** [comment]
> If helpful (x86_64 as by `uname -m`) for Debian 12/bookworm (testing) with OpenBabel 3.1.1.+dfsg-6, the generation of the thiazole .sdf passes without complain or error. Despite the presence of the two bridged bicyclic structures, both for `--gen2d` and alternative `--gen3d`.

I believe this issue is specific to arm64 architecture.

**Comment by fredrikw** [comment]
Sorry for the spamming, I don't have access to an arm64 computer so I will ask you to do a couple of more things. Could you please check
`obabel -ismi -:"C1CCCCC1" -osdf --gen2D` and `obabel -ismi -:"c1ccccc1" -osdf --gen2D` and `obabel -ismi -:"c1ccncc1" -osdf --gen2D`

**Comment by merkys** [comment]
> Sorry for the spamming, I don't have access to an arm64 computer so I will ask you to do a couple of more things. Could you please check `obabel -ismi -:"C1CCCCC1" -osdf --gen2D` and `obabel -ismi -:"c1ccccc1" -osdf --gen2D` and `obabel -ismi -:"c1ccncc1" -osdf --gen2D`

Just checked, all of these work, and produce meaningful looking SDF output. Let me know should you want me to check anything else on arm64.

**Comment by fredrikw** [comment]
Are you in a position to debug the build? In that case I would like to see what's happening around https://github.com/openbabel/openbabel/blob/f3ed2a9a5166dbd3b9ce386e636a176074a6c34c/src/mcdlutil.cpp#L2575
In particular, the content of dsATN and the values of nb and i when we get the error.

**Comment by merkys** [comment]
I tried both `break mcdlutil.cpp:2575` and `break src/mcdlutil.cpp:2575`, but every time I get
```
No source file named mcdlutil.cpp.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (mcdlutil.cpp:2575) pending.
```
And then doing `run` just proceeds until the end of the program, and I get the same output as without the breakpoint. I guess I will need to look through some tutorial on gdb usage.

**Comment by merkys** [comment]
OK, I managed to get the breakpoint working. Will let you know if I dig something out.

**Comment by Crivella** [comment]
This is the exact same problem that was popping out on

- https://github.com/openbabel/openbabel/issues/2138

In that case for x86 it only happens when compiling with `-march=native`.
In that PR it was dismissed as a compiler quirk...
I assume the default behavior of GCC on arm will be different, but error dependent on optimization usually pops up due to using undefined code behavior.

**Comment by Crivella** [comment]
Using the same command as @merkys 

```
gdb --args obabel -ismi -:"O=C(C1=CN=CS1)N1C2CCC1CN(CC1CC3CCC1O3)C2" -osdf --gen2D 

(gdb) run
```

was able to pin down the failure at

https://github.com/openbabel/openbabel/blob/889c350feb179b43aa43985799910149d4eaa2bc/src/mcdlutil.cpp#L4290

Where `this->getAtom(naDef)->ac[n]` evaluates to 16 while nAtoms is 8

<details><summary>BACKTRACE</summary>
<p>

(Had to remove the part of the code that was catching the exception and giving a very non descriptive error instead)

```
#0  0x00004000000fe570 in ?? () from /cvmfs/software.eessi.io/versions/2023.06/compat/linux/aarch64/lib/../lib64/libc.so.6
#1  0x00004000000b96bc in raise () from /cvmfs/software.eessi.io/versions/2023.06/compat/linux/aarch64/lib/../lib64/libc.so.6
#2  0x00004000000a67bc in abort () from /cvmfs/software.eessi.io/versions/2023.06/compat/linux/aarch64/lib/../lib64/libc.so.6
#3  0x0000400000684c48 in __gnu_cxx::__verbose_terminate_handler () at /tmp/bot/easybuild/build/GCCcore/12.3.0/system-system/gcc-12.3.0/stage3_obj/aarch64-unknown-linux-gnu/libstdc++-v3/libsupc++/vterminate.cc:95
#4  0x000040000068266c in __cxxabiv1::__terminate (handler=<optimized out>) at /tmp/bot/easybuild/build/GCCcore/12.3.0/system-system/gcc-12.3.0/stage3_obj/aarch64-unknown-linux-gnu/libstdc++-v3/libsupc++/eh_terminate.cc:48
#5  0x00004000006826d0 in std::terminate () at /tmp/bot/easybuild/build/GCCcore/12.3.0/system-system/gcc-12.3.0/stage3_obj/aarch64-unknown-linux-gnu/libstdc++-v3/libsupc++/eh_terminate.cc:58
#6  0x00004000006829b0 in __cxxabiv1::__cxa_throw (obj=<optimized out>, tinfo=0x4000007e7ee0 <typeinfo for std::out_of_range>, dest=0x400000697230 <std::out_of_range::~out_of_range()>)
    at /tmp/bot/easybuild/build/GCCcore/12.3.0/system-system/gcc-12.3.0/stage3_obj/aarch64-unknown-linux-gnu/libstdc++-v3/libsupc++/eh_throw.cc:98
#7  0x00004000006ac538 in std::__throw_out_of_range_fmt (__fmt=<optimized out>) at /tmp/bot/easybuild/build/GCCcore/12.3.0/system-system/gcc-12.3.0/stage3_obj/aarch64-unknown-linux-gnu/libstdc++-v3/src/c++11/functexcept.cc:101
#8  0x0000400000303924 in OpenBabel::TSimpleMolecule::getAtom(int) () from /home/crivella/easybuild/128686/OpenBabel/3.1.1/gompi-2023a/easybuild_obj/bin/../lib/libopenbabel.so.7
#9  0x00004000003165a0 in OpenBabel::TEditedMolecule::addAsTemplate(OpenBabel::TEditedMolecule&, int, int, int, int, bool) () from /home/crivella/easybuild/128686/OpenBabel/3.1.1/gompi-2023a/easybuild_obj/bin/../lib/libopenbabel.so.7
#10 0x000040000031fc4c in OpenBabel::TemplateRedraw::coordinatesPrepare(OpenBabel::TEditedMolecule&, int, int) () from /home/crivella/easybuild/128686/OpenBabel/3.1.1/gompi-2023a/easybuild_obj/bin/../lib/libopenbabel.so.7
#11 0x00004000003271b4 in OpenBabel::TemplateRedraw::redrawFine(OpenBabel::TSimpleMolecule&) () from /home/crivella/easybuild/128686/OpenBabel/3.1.1/gompi-2023a/easybuild_obj/bin/../lib/libopenbabel.so.7
#12 0x0000400000328780 in OpenBabel::generateDiagram(OpenBabel::OBMol*) () from /home/crivella/easybuild/128686/OpenBabel/3.1.1/gompi-2023a/easybuild_obj/bin/../lib/libopenbabel.so.7
#13 0x00004000019da884 in OpenBabel::OpGen2D::Do(OpenBabel::OBBase*, char const*, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const*, OpenBabel::OBConversion*) () from /home/crivella/easybuild/128686/OpenBabel/3.1.1/gompi-2023a/easybuild_obj/lib/plugin_ops.so
#14 0x000040000038d31c in OpenBabel::OBMol::DoTransformations(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const*, OpenBabel::OBConversion*) () from /home/crivella/easybuild/128686/OpenBabel/3.1.1/gompi-2023a/easybuild_obj/bin/../lib/libopenbabel.so.7
#15 0x00004000003968a0 in OpenBabel::OBMoleculeFormat::ReadChemObjectImpl(OpenBabel::OBConversion*, OpenBabel::OBFormat*) () from /home/crivella/easybuild/128686/OpenBabel/3.1.1/gompi-2023a/easybuild_obj/bin/../lib/libopenbabel.so.7
#16 0x0000400000342884 in OpenBabel::OBConversion::Convert() () from /home/crivella/easybuild/128686/OpenBabel/3.1.1/gompi-2023a/easybuild_obj/bin/../lib/libopenbabel.so.7
#17 0x0000400000348c88 in OpenBabel::OBConversion::Convert(std::istream*, std::ostream*) () from /home/crivella/easybuild/128686/OpenBabel/3.1.1/gompi-2023a/easybuild_obj/bin/../lib/libopenbabel.so.7
#18 0x000040000034a600 in OpenBabel::OBConversion::FullConvert(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&) () from /home/crivella/easybuild/128686/OpenBabel/3.1.1/gompi-2023a/easybuild_obj/bin/../lib/libopenbabel.so.7
#19 0x0000000000403860 in main ()
```

</p>
</details> 

Might try to dig more but I think having someone that knows what the code is doing looking at this might actually helpful

**Comment by Crivella** [comment]
Got it!

Usage of uninitialized arrays leading to undefined behavior...
https://github.com/openbabel/openbabel/pull/2851

EDIT: (at least the undefined behavior causing different opt arch to five this error)

I think the entire section
https://github.com/openbabel/openbabel/blob/889c350feb179b43aa43985799910149d4eaa2bc/src/mcdlutil.cpp#L4279-L4293
might need better guards/checks as somehow i think
```
          if(r2 < amin) {
            amin=r2;
            n=i;
          };
```
is being skipped (possibly due to NaN's? I saw some when debugging) causing the following section to check for `n=1` even when `nb` is 1 (which should make only `->ac[0]` valid as far as i understand).
Alternatively not sure if `nb` should just not be 1 for this system, but i will let the devs check the code logic ;)

**Comment by ghutchis** [comment]
Thanks to @Crivella for a bunch of patches on this. I'm adding a Linux-ARM builder with #2898 - I'll continue to look through mcdutil over the next week.

**Comment by ghutchis** [comment]
Everything seems okay now on ARM so I'm going to close this. If something else pops up, please open a new issue.
